### DISCUSSION TITLE
Typescript error when dispatching thunks

### DISCUSSION THREAD

**Discussion by omniphx** [reduxjs/redux-toolkit#4090] [Q&A]
I'm in the process of upgrading our codebase to RTK v2.0 and running into an Typescript error when I try to pass my thunk into a dispatch method. Looks like dispatch only accepts `UnknownAction`. Is there a different way to implement thunks now?

See the comments at the bottom for the error I'm getting:
```typescript
import { configureStore, createAsyncThunk } from '@reduxjs/toolkit';

const store = configureStore({
  reducer: {},
});

interface MyData {
  h: string;
}

const fetchUserById = createAsyncThunk(
  'users/fetchById',
  // Declare the type your function argument here:
  async (userId: number) => {
    const response = await fetch(`https://reqres.in/api/users/${userId}`);
    // Inferred return type: Promise<MyData>
    return (await response.json()) as MyData;
  },
);

const request = async () => {
  // Typescript error: `(property) dispatch: Dispatch<UnknownAction>(action: UnknownAction, ...extraArgs: any[]) => UnknownAction`
  const result = await store.dispatch(fetchUserById(3));
};
```

Edit: Here's a StackBlitz example: https://stackblitz.com/edit/react-ts-hmoqp3?file=App.tsx

**Comment by EskiMojo14** [comment]
Any chance you could get a reliable reproduction of this in either Typescript Playground or CodeSandbox (or a Github repo)? I've seen this crop up a couple of times, but not reliably enough to work out what's causing it.

**Reply by markerikson** [reply]
Also, while it's not directly related, I'm nervous about that `store.dispatch` - generally you shouldn't be importing or referencing the store directly.  Is that just an example of "this is what happens when I dispatch _at all_"?  Or are you actually using `store.dispatch` in the code?

(This also matters because if you're actually doing, say, `const dispatch = useDispatch()` and not `store.dispatch`, that comes into play also.)

**Reply by omniphx** [reply]
@markerikson, correct trying to reproduce the typescript error as an example of what's happening. In our codebase, we are using useDispatch().

^ @EskiMojo14 I'll try to get an example in a code sandbox

**Reply by EskiMojo14** [reply]
when you say you're using useDispatch, have you followed the guide to get [typed versions of your hooks](https://react-redux.js.org/using-react-redux/usage-with-typescript#define-typed-hooks)?

**Reply by omniphx** [reply]
@EskiMojo14 yup using the typed version of `useDispatch`. Here's the sandbox example: https://stackblitz.com/edit/react-ts-hmoqp3?file=App.tsx

**Reply by markerikson** [reply]
Hmm. Definitely something wrong there. `AppDispatch` isn't getting any of the thunk typing inferred:

![image](https://github.com/reduxjs/redux-toolkit/assets/1128784/a866f27c-1848-4bb3-a342-a6b5e1ccf6ff)

**Reply by markerikson** [reply]
I will say that the `react-redux` dep is definitely wrong - you shouldn't be specifying multiple versions like this:

```ts
"react-redux": "^7.2.1 || ^8.1.3 || ^9.0.0",
```

Just use `"^9"`, one version.

You also don't need to list `redux` separately, although in this case it shouldn't be making a difference.

But yeah, not immediately sure why this is getting inferred wrong.

**Reply by EskiMojo14** [reply]
huh, I'm not seeing that issue in the link - the only issues i can see is that your store needs at least one reducer and that you're not actually calling the thunk creator (i.e. should be dispatch(getUserById(1)) not dispatch(getUserById))

**Reply by omniphx** [reply]
@EskiMojo14 Just trying to recreate the Typescript error. Here's a screenshot:
<img width="858" alt="Screenshot 2024-01-17 at 5 42 04 PM" src="https://github.com/reduxjs/redux-toolkit/assets/3722405/f98bad73-b9de-4eba-836d-85701564eae9">

@markerikson The stackblitz tool automatically added it that way for some reason ¯\_(ツ)_/¯ Went ahead and changed it to `^9`

Appreciate the help on this!

**Reply by EskiMojo14** [reply]
yeah, that type error is because you're trying to dispatch a thunk *creator* not a thunk.

**Comment by omniphx** [comment]
Ended up discovering that our codebase has a library `@types/redux-mock-store` that was causing `redux` to resolve to down to v4. Not sure why there was a similar error in the stackblitz example. Seems to just be a coincidence/red herring with something happening behind the scenes. Even tried creating a CRA app with the same exact code and had no Typescript issues.

@markerikson @EskiMojo14 thanks again for helping out. Just wanted to document my findings in case anyone else runs into something similar.
